### PR TITLE
Add mrr calculation

### PR DESCRIPTION
**Pull request by zkid18**
## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [x] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [x] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [x] Did you check the docs with `make check-docs`?
- [x] Did you write any new necessary tests?
- [x] Did you add your new functionality to the docs?
- [x] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/CHANGELOG.md)?
- [x] **You can use 'Login as guest' to see Teamcity build logs.**

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: catalyst/utils/metrics/mrr.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import torch
+
+
+def mrr(outputs: torch.Tensor, targets: torch.Tensor):
```
Comment (Scitator) [inline catalyst/utils/metrics/mrr.py]: could you please add docs here
https://github.com/catalyst-team/catalyst/blob/master/docs/api/utils.rst
?

Comment (Scitator) [reply catalyst/utils/metrics/mrr.py]: btw, what do you think about MRRCallback? like https://github.com/catalyst-team/catalyst/blob/master/catalyst/dl/callbacks/metrics/iou.py#L12 for example

**Comment by zkid18** [comment]
I still consider the proper design for MRR@K. Probably `_metric@k` is quite common task, so rather than implement this functionality inside every metrics extend _metric with, say, `topKMixin`

File: catalyst/utils/metrics/mrr.py (hunk 1 above)
Comment (zkid18) [reply catalyst/utils/metrics/mrr.py]: Thanks, I'll have a look at callbacks.

File: catalyst/utils/metrics/mrr.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    """
+    Calculate the MRR score given model ouptputs and targets
+    Args:
+        outputs [batch_size, slate_length] (torch.Tensor): model outputs, logits
+        targets [batch_szie, slate_length] (torch.Tensor): ground truth, labels
+    Returns:
+        mrr (float): the mrr score
+    """
+    outputs = outputs.clone()
+    targets = targets.clone()
```
Comment (Scitator) [inline catalyst/utils/metrics/mrr.py]: why do you need `clone`?

Comment (zkid18) [reply catalyst/utils/metrics/mrr.py]: Tried to follow the  'shared clones', which is a common pattern in torch. But it seems it not so necessary here.

File: catalyst/utils/metrics/mrr.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
+"""
+MRR metric.
+"""
+
+import torch
+
+
+def mrr(outputs: torch.Tensor,
```
Comment (Scitator) [inline catalyst/utils/metrics/mrr.py]: could you please add this metric to docs?
https://github.com/catalyst-team/catalyst/blob/master/docs/api/utils.rst#metrics

File: catalyst/dl/callbacks/metrics/mrr.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
+from typing import List
+
+from catalyst.core import MetricCallback
+from catalyst.utils import metrics
+
+
+class MRRCallback(MetricCallback):
```
Comment (Scitator) [inline catalyst/dl/callbacks/metrics/mrr.py:5]: could you please add this callback to the docs?
https://github.com/catalyst-team/catalyst/blob/master/docs/api/dl.rst#metrics

**Review by Scitator — DISMISSED** [review]
btw, looks like we need to fix codestyle

File: catalyst/dl/callbacks/metrics/mrr.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        activation: str = "none",
+    ):
+        """
+        Args:
+            input_key (str): input key to use for auc calculation
+                specifies our ``y_true``
+            output_key (str): output key to use for auc calculation;
+                specifies our ``y_pred``
+            prefix (str): name to display for mrr when printing
+            activation (str): An torch.nn activation applied to the outputs.
```
Comment (Scitator) [inline catalyst/dl/callbacks/metrics/mrr.py]: why do we need `activation`? mrr metric doesn't have `activation` support

**Review by Scitator — DISMISSED** [review]

**Comment by Scitator** [comment]
@zkid18 could you please add a sanity check test for MRRCallback? like "init and compute on one sample"

**Comment by zkid18** [comment]
> @zkid18 could you please add a sanity check test for MRRCallback? like "init and compute on one sample"

Should I write tests for that? Didn't find a dedicated place for that

**Comment by Scitator** [comment]
@zkid18 You could place callback tests to 
```
catalyst/dl/callbacks/metrics/tests/test_mrr.py
```
:)

File: catalyst/utils/metrics/__init__.py [hunk 6]
```diff
@@ -3,10 +3,12 @@
     accuracy,
     multi_label_accuracy,
 )
+
 from catalyst.utils.metrics.auc import auc
 from catalyst.utils.metrics.cmc_score import cmc_score_count, cmc_score
 from catalyst.utils.metrics.dice import dice, calculate_dice
 from catalyst.utils.metrics.f1_score import f1_score
+from catalyst.utils.metrics.mrr import mrr
```
Comment (bagxi) [inline catalyst/utils/metrics/__init__.py]: could you please follow the alphabetical order of imports?

File: catalyst/dl/callbacks/metrics/tests/test_mrr.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    model=model,
+    criterion=criterion,
+    optimizer=optimizer,
+    scheduler=scheduler,
+    loaders=loaders,
+    logdir="./logdir",
+    num_epochs=2,
+    verbose=True,
+    callbacks=[MRRCallback, SchedulerCallback(reduced_metric="loss")]
+)
```
Comment (bagxi) [inline catalyst/dl/callbacks/metrics/tests/test_mrr.py]: new line at the end of py file is missed

File: catalyst/utils/metrics/mrr.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
+"""
+MRR metric.
+"""
+
+import torch
+
+
+def mrr(outputs: torch.Tensor, targets: torch.Tensor, k=100) -> torch.Tensor:
```
Comment (bagxi) [inline catalyst/utils/metrics/mrr.py]: could you please define the types for all args?

File: catalyst/utils/metrics/mrr.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+            for the user and 0 not relevant
+            size: [batch_szie, slate_length]
+            ground truth, labels
+
+    Returns:
+        result (torch.Tensor): 
+            The mrr score for each user.
+            
+    """
+    k = min(outputs.size()[1], k)
```
Comment (bagxi) [inline catalyst/utils/metrics/mrr.py]:
```suggestion
    k = min(outputs.size(1), k)
```

File: catalyst/utils/metrics/mrr.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+MRR metric.
+"""
+
+import torch
+
+
+def mrr(outputs: torch.Tensor, targets: torch.Tensor, k=100) -> torch.Tensor:
+
+    """
+    Calculate the MRR score given model ouptputs and targets
```
Comment (bagxi) [inline catalyst/utils/metrics/mrr.py]: I'd appreciate it if you could extend documentation with the explanation of the metric or add link users will be able to read more about this score.

File: catalyst/utils/metrics/mrr.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+            size: [batch_szie, slate_length]
+            ground truth, labels
+
+    Returns:
+        result (torch.Tensor): 
+            The mrr score for each user.
+            
+    """
+    k = min(outputs.size()[1], k)
+    _, indices_for_sort = outputs.sort(descending=True, dim=-1)
```
Comment (bagxi) [inline catalyst/utils/metrics/mrr.py]: could we use [torch.topk](https://pytorch.org/docs/stable/generated/torch.topk.html#torch.topk) here?

**Review by bagxi — DISMISSED** [review]

File: catalyst/utils/metrics/mrr.py (hunk 11 above)
Comment (zkid18) [reply catalyst/utils/metrics/mrr.py]: I guess the comment is more relevant to the next part
`true_sorted_by_pred_shrink = true_sorted_by_preds[:, :k]`

Anyway, I might missing the advantage of `torch.topk` over the proposed approach. 
We need to sort predictions by the corresponding indexes of the outputs. Is there is a way in pytorch to sort in that fashion?

File: catalyst/dl/callbacks/metrics/mrr.py (hunk 4 above)
Comment (zkid18) [reply catalyst/dl/callbacks/metrics/mrr.py:5]: As for tests, I think better return to the question when we implement at least one Learning to Rank models.

File: CHANGELOG.md [hunk 12]
```diff
@@ -39,7 +39,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 ## [20.08] - 2020-08-09
 
 ### Added
-- Full metric learning pipeline including training and validation stages ([#886](https://github.com/catalyst-team/catalyst/pull/876))
+- MRR metrics calculation ([#886](https://github.com/catalyst-team/catalyst/pull/886))
```
Comment (Scitator) [inline CHANGELOG.md]: looks like we need to move it up :)

**Review by Scitator — DISMISSED** [review]

**Comment by Scitator** [comment]
@zkid18 could you please update your branch and resolver the above questions? thank you!

**Comment by zkid18** [comment]
@Scitator have fixed the typos, and moved `mrr` metric under `metrics` folder. 
Could you check if I understood this idea correctly?

**Comment by zkid18** [comment]
@Scitator 
Not sure if this problem related to codestyle itself or rather actions pipeline. 
https://github.com/catalyst-team/catalyst/runs/1224097493?check_suite_focus=true

File: docs/api/utils.rst [hunk 13]
```diff
@@ -290,3 +290,10 @@ Functional
     :members:
     :undoc-members:
     :show-inheritance:
+
+MRR
+~~~~~~~~~~~~~~~~~~~~~~
+.. automodule:: catalyst.utils.metrics.mrr
```
Comment (Scitator) [inline docs/api/utils.rst]: looks like here is an error with docs :)

**Review by Scitator — DISMISSED** [review]
